submissions/factorio_draftsman/main.md

```diff
@@ -0,0 +1,200 @@
+
+## Automatic Item Stack Sizes
+
+Oftentimes for you want to figure out how much storage you need for an amount of an item. However, storage in Factorio is based on slots, not amounts, so the amount of storage is actually dependent on not only the amount you're trying to store but also the stack size of that item. [You can design a circuit contraption](https://factorioprints.com/view/-LwH68dlvOEIKWkU5JqR) to determine this amount of slots that you need, but you need a big cell of combinators documenting the item sizes. This is *boring* to make, and easily broken if a new item is added or the stack size changes. Clearly, for such a simple and repetitive task, a script is better suited:
+
+```python
+# Create a N x 5 grid of connected constant combinators, with every item and their stack size
+
+from draftsman.blueprintable import Blueprint
+from draftsman.constants import Direction
+from draftsman.data import items
+from draftsman.entity import ConstantCombinator
+
+
+def main():
+    blueprint = Blueprint()
+
+    count = 0  # Total number of signals
+    index = 0  # Signal index in the current combinator
+    i = 0  # How many combinators we've added
+    x = 0
+    y = 0
+    combinator = ConstantCombinator(direction=Direction.SOUTH)
+
+    # Iterate over every item in order:
+    for item in items.raw:
+        # Ignore hidden items/entities
+        if "flags" in items.raw[item]:
+            if "hidden" in items.raw[item]["flags"]:
+                continue
+        # Keep track of how many signals we've gone through
+        count += 1
+        # Write the stack size signal
+        stack_size = items.raw[item]["stack_size"]
+        combinator.set_signal(index, item, stack_size)
+        index += 1
+        # Once we exceed the current combinator, place it and reset
+        if index == 20:
+            # Add the combinator to the blueprint
+            combinator.id = "{}_{}".format(x, y)
+            blueprint.entities.append(combinator)
+            # Reset the combinator
+            i += 1
+            y = i % 5
+            x = int(i / 5)
+            combinator.set_signals(None)  # Clear signals
+            combinator.tile_position = (x, y)
+            index = 0
+
+    # Add the last combinator if partially full
+    combinator.id = "{}_{}".format(x, y)
+    blueprint.entities.append(combinator)
+
+    # Add connections to each neighbour
+    for cx in range(x):
+        for cy in range(5):
+            here = "{}_{}".format(cx, cy)
+            right = "{}_{}".format(cx + 1, cy)
+            below = "{}_{}".format(cx, cy + 1)
+            try:
```

Comment:
You need the try-excepts because you're always connecting columns of combinators at a time without taking into account that the last combinator isn't necessarily at the bottom of its column,, right? I'd add this instead:
```py
if cx == x && cy > c: continue
```

Reply:
Well, more so that each combinator is supposed to be connected to its right and below neighbor in order to create a regular grid of wires. However, if it's on the rightmost column, it will attempt to connect right to an entity that's not there, and on the bottom row, connect below to an entity that's not there, hence the `KeyError`s. `KeyError` in this case could be renamed to `MissingEntityError`, which would be more specific. I could iterate by `x - 1`, by `5 - 1` instead, but that wouldn't complete all the connections on the right column or bottom row, because half of the connections on those rows are valid. Instead of messing around with complex conditions or multiple loop passes, I found the simplest way to connect them was to simply iterate over a grid of the same size and use Python's `try` to attempt the connection if possible and `pass` (ignore the error) if it happens to fails on the edges.

I admit though that this is ugly, and to this day I've never been a fan of the way it looks. I could do a similar method without of checking using the try catch, and explicitly check to see if the entity exists before attempting the connection; would this be clearer?
```python
for cx in range(x):
    for cy in range(COMBINATOR_HEIGHT):
        here = "{}_{}".format(cx, cy)
        right = "{}_{}".format(cx + 1, cy)
        below = "{}_{}".format(cx, cy + 1)

        # get() in this case would attempt to get the value of key right, and return False if not found
        if blueprint.entities.get(right, False): 
            blueprint.add_circuit_connection("red", here, right)
        if blueprint.entities.get(below, False):
            blueprint.add_circuit_connection("red", here, below)
```
*Technically* this isn't implemented yet, so I guess this is also an API question. `get()` is commonly used on dictionary-like objects, of which my `EntityList` has functionality from. Alternatively, you could also have something like `blueprint.contains(string_id)`.

All other of your proposed changes have been resolved.

Reply:
I see, I had the wrong idea of what the loop was doing, thank you for elaborating!

Keep the try-excepts you have now then, Ask Forgiveness Not Permission and all being the Python mantra for clean code. :)